src/pages/resultsView/ResultsViewPageStore.ts

```diff
@@ -1123,40 +1132,96 @@ export class ResultsViewPageStore {
+                for (const sampleSpec of this._sampleSpecification.result!) {
+                    if (sampleSpec.sampleId) {
+                        ret[sampleSpec.studyId] = ret[sampleSpec.studyId] || {};
+                        ret[sampleSpec.studyId][sampleSpec.sampleId] = true;
+                    } else if (sampleSpec.sampleListId) {
+                        sampleListsToQuery.push(sampleSpec as { studyId: string, sampleListId: string });
+                    }
+                }
+                const results: string[][] = await Promise.all(sampleListsToQuery.map(spec => {
+                    return client.getAllSampleIdsInSampleListUsingGET({
+                        sampleListId: spec.sampleListId
+                    });
+                }));
+                for (let i = 0; i < results.length; i++) {
+                    ret[sampleListsToQuery[i].studyId] = ret[sampleListsToQuery[i].studyId] || {};
+                    const sampleMap = ret[sampleListsToQuery[i].studyId];
+                    results[i].map(sampleId => {
+                        sampleMap[sampleId] = true;
+                    });
+                }
+                return ret;
         }
-        return ret;
+
     }, {});
 
-    @computed get studyToSampleListId(): { [studyId: string]: string } {
-        return this.samplesSpecification.reduce((map, next) => {
-            if (next.sampleListId) {
-                map[next.studyId] = next.sampleListId;
-            }
-            return map;
-        }, {} as {[studyId: string]: string});
-    }
+    // @computed get studyToSampleListId(): { [studyId: string]: string } {
+    //     return this.samplesSpecification.reduce((map, next) => {
+    //         if (next.sampleListId) {
+    //             map[next.studyId] = next.sampleListId;
+    //         }
+    //         return map;
+    //     }, {} as {[studyId: string]: string});
+    // }
+
+    readonly studyToSampleListId = remoteData<{ [studyId: string]: string }>({
+        await:()=>[
+            this._sampleSpecification
+        ],
+        invoke: async ()=>{
+            return this._sampleSpecification.result!.reduce((map, next) => {
+                if (next.sampleListId) {
+                    map[next.studyId] = next.sampleListId;
+                }
+                return map;
+            }, {} as {[studyId: string]: string});
+        }
+    })
+
+
+    readonly _sampleSpecification = remoteData({
+       invoke: async ()=>{
+
+           // is this a sample list category query?
+           // if YES, we need to derive the sample lists by:
+           // 1. looking up all sample lists in selected studies
+           // 2. using those with matching category
+           if (!this.sampleListCategory) {
+               return this.samplesSpecification
+           } else {
+               const allSampleLists = await Promise.all(this.samplesSpecification.map((spec)=>{
+                   return client.getAllSampleListsInStudyUsingGET({
```

Comment:
your loop on line 1208+ could end up taking quite a long time this way if theres a long custom case list. instead you could collect the unique study ids then query for those

Reply:
it's true but we know that if there's a category at play, then there can only ever be as many spec elements as studies and they will be unique

Reply:
ah okay, fair enough. maybe add a comment about this?